### ISSUE TITLE
Semantic Segmentation Project - Improvements & Issues Tracker

### ISSUE THREAD

**Issue by fra2404** [fra2404/segmentation-cityscape#1]
## 📋 Project Overview
Semantic segmentation implementation using DeepLabV3-ResNet101 on Cityscapes dataset with PyTorch and MPS acceleration.

### 1. **Test Set Evaluation Returns Zero Metrics**
- **Status**: Open
- **Priority**: High
- **Location**: Cell [13]
- **Problem**: Test set shows 0.0 accuracy and IoU
- **Root Cause**: Cityscapes test set lacks local ground truth annotations
- **Solution**: 
  - Use validation set for quantitative metrics
  - Add clear warning about test set limitations
  - Implement test inference-only pipeline

### 2. **Poor Model Performance (Low mIoU: 0.074)**
- **Status**: Open
- **Priority**: High
- **Location**: Evaluation Results
- **Problem**: Mean IoU is very low (7.4%)
- **Possible Causes**:
  - Insufficient training (only 3 epochs)
  - Severe class imbalance
  - Suboptimal hyperparameters
- **Solutions**:
  - Increase training epochs to 20-30
  - Implement focal loss
  - Tune learning rate scheduler
  - Add more data augmentation

### 3. **Class Imbalance Issues**
- **Status**: Open
- **Priority**: Medium
- **Location**: Class Balance Analysis
- **Problem**: Some classes have <0.5% frequency
- **Current**: Dynamic class weights calculated
- **Improvements Needed**:
  - Implement focal loss (gamma=2.0)
  - Add class-weighted sampling
  - Apply class weight clamping

### 4. **Training Configuration Optimization**
- **Status**: Open
- **Priority**: Medium
- **Location**: Training Setup
- **Issues**:
  - Small batch size (2-4) due to memory constraints
  - Only 3 epochs in demo run
  - Basic learning rate scheduler
- **Proposed Changes**:
  ```python
  # Increase effective training
  NUM_EPOCHS = 20
  BATCH_SIZE = 8 (with gradient accumulation)
  ```

 ### 5. Data Augmentation Enhancement
- **Status**: Open
- **Priority**: Medium
- **Location**: Data Transforms
- **Current**: Basic augmentation (flip, color jitter, affine)
- **Proposed Enhancements**:
```python
Add Gaussian blur
Add random grayscale
Add cutout/mixup
Increase resolution to 512x512
 ```

### 6. MPS Optimization

- **Status**: Open
- **Priority**: Low
- **Location**: Throughout notebook
- **Issues**:
 - Potential MPS compatibility issues
 - Slow training on macOS
- Solutions:
 - Optimize dataloader for MPS (num_workers=0)
 - Add MPS-specific memory management
 - Implement gradient checkpointing

### 7. Code Organization & Scripting
- **Status**: Open
- **Priority**: Medium
- **Issues**:
 - All code in single notebook
 - No modular structure
 - Hard to reproduce
 - Proposed Structure:

```text
src/
├── data/
│   ├── dataset.py
│   ├── transforms.py
│   └── preprocessing.py
├── models/
│   ├── deeplabv3.py
│   └── losses.py
├── training/
│   ├── trainer.py
│   └── metrics.py
├── utils/
│   ├── visualization.py
│   └── logging.py
└── configs/
    └── default.yaml
```

### 8. Evaluation Pipeline Improvement

- **Status**: Open
- **Priority**: Medium
- **Issues**:
 - Basic IoU calculation
 - No per-class metrics visualization
 - No confusion matrix
- Additions Needed:
 - Precision, Recall, F1-score per class
 - Confusion matrix visualization
 - Learning curves with TensorBoard/W&B
 - Model comparison capabilities

 
### 9. Test Set Submission PipelineStatus: OpenPriority: Low
- **Current**: Basic prediction saving
- **Improvements**:
 - Automated ZIP creation for Cityscapes server
 - Submission format validation
 - Results parsing from server response
